pkg/ifuzz/powerpc/powerpc.go

```diff
@@ -51,16 +50,26 @@ func (insnset *InsnSet) GetInsns(mode iset.Mode, typ iset.Type) []iset.Insn {
 }
 
 func (insnset *InsnSet) Decode(mode iset.Mode, text []byte) (int, error) {
-	if len(text) < 4 {
-		return 0, errors.New("must be at least 4 bytes")
-	}
-	insn32 := binary.LittleEndian.Uint32(text)
-	for _, ins := range insnset.Insns {
-		if ins.Mask&insn32 == ins.Opcode {
-			return 4, nil
+	i := 0
+	for ; i < len(text); i += 4 {
```

Comment:
Hummm... x86 decoder does not do this, right?
Which makes me think if x86 needs to do the same? I think x86 pseudo instructions can also produce multiple real instructions. Or if the intention was actually to handle multiple instructions in the callers of this function?
I see here split calls Decode repeatedly to decode multiple instructions:
https://github.com/google/syzkaller/blob/df4265465adfccfb96d35b0078d73bce8e16e4d7/pkg/ifuzz/ifuzz.go#L120-
Is this really necessary? Is there some other caller that doesn't handle multiple instructions? Perhaps it needs to be switched to use of the split function?
Handling only 1 instruction in per-arch code and handling multiple in the common code looks reasonable from code structuring point of view.

Reply:
Ah you're right, I did this before I figured what else was broken, and I also thought at some point that the decoder must see the instruction which has just been encoded. The only reason for thus hunk was to pass the tests. I'll revert this.

Reply:
The easiest thing seems to be hacking `pkg/ifuzz/ifuzz_test.go` to treat all instructions as pseudo and `goto repeat` if there are still instructions to decode in `text`, no?

Reply:
I think the precise size check is a good cross-check for decoder correctness (otherwise it can always say each byte is an instruction and everything will work). So I hate to lose this check for x86.
The test takes pseudo from Info:
```
				name, _, pseudo, _ := insn.Info()
```
You can make ppc Info always return true.

Reply:
This may have other unwanted complications in places relying on this `insn.Info()`. Which only seems to affect `TypeExec` which I forgot - what is it for? The name is not self explanatory :(

Reply:
I think the intended meaning is this:
https://github.com/google/syzkaller/blob/94e0b7079acc1c1ab28a2043fd21321c978f1529/pkg/ifuzz/iset/iset.go#L39

Say, if we would test objdump disassembler with ifuzz output, then we don't want to use instructions that are only interesting from "execution" point of view (it's assumed that all individual instructions produced by pseudo instructions can be also generated as normal instructions). So using pseudo instructions for disassembler testing would just significantly increase presence of IN/OUT/RDMSR/WRMSR instructions.

But I don't think we currently use Exec=false anywhere, at least prog package sets Exec=true.

An alternative would be add a flag to iset saying that Decode needs to be called multiple times for all instructions. Then you can enable it for ppc.

Reply:
for now I look at MemRegions and do not generate extra instructions if none in the cfg (which is the ifuzz_test's case)

Reply:
btw where is that "disassembler testing" and objdump happening? is it that ifuzz_test only or there is more (which I missed and should probably fix)?

Reply:
nowhere
I meant theoretically ifuzz can be used to test/fuzz objdump.